server/src/delta_cache.rs

```diff
@@ -0,0 +1,286 @@
+        },
+        DeltaEvent::SegmentUpdated {
+            event_id: 5,
+            segment: Segment { id: 1,  constraints: vec![] },
+        },
+        DeltaEvent::SegmentRemoved {
+            event_id: 6,
+            segment_id: 2,
+        },
+        DeltaEvent::SegmentUpdated {
+            event_id: 7,
+            segment: Segment { id: 3, constraints: vec![] },
+        },
+    ];
+    delta_cache.add_events(added_events.clone());
+
+    let events: Vec<_> = delta_cache.get_events().iter().cloned().collect();
+    assert_eq!(events.len(), max_length);
+    assert_eq!(events, added_events[added_events.len() - max_length..]);
+
+    let hydration_event = delta_cache.get_hydration_event();
+    assert_eq!(hydration_event.features.len(), 2);
+    assert_eq!(hydration_event.event_id, 7);
+    assert!(hydration_event.features.iter().any(|f| f.name == "my-feature-flag"));
+    assert!(hydration_event.features.iter().any(|f| f.name == "another-feature-flag"));
+    assert!(hydration_event.segments.iter().any(|s| s.id == 1));
+}
+
+#[test]
+fn test_prevent_mutation_of_previous_feature_updated_events() {
+    let base_event = DeltaHydrationEvent {
+        event_id: 1,
+        features: vec![ClientFeature {
+            name: "base-flag".to_string(),
+            feature_type: None,
+            description: None,
+            created_at: None,
+            last_seen_at: None,
+            enabled: true,
+            stale: None,
+            impression_data: None,
+            project: None,
+            strategies: None,
+            variants: None,
+            dependencies: None,
+        }],
+        segments: vec![],
+    };
+    let mut delta_cache = DeltaCache::new(base_event, 10);
+
+    let initial_feature_event = DeltaEvent::FeatureUpdated {
+        event_id: 129,
+        feature: ClientFeature {
+            name: "streaming-test".to_string(),
+            feature_type: None,
+            description: None,
+            created_at: None,
+            last_seen_at: None,
+            enabled: false,
+            stale: None,
+            impression_data: None,
+            project: None,
+            strategies: None,
+            variants: None,
+            dependencies: None,
+        },
+    };
+    delta_cache.add_events(vec![initial_feature_event.clone()]);
+
+    let updated_feature_event = DeltaEvent::FeatureUpdated {
+        event_id: 130,
+        feature: ClientFeature {
+            name: "streaming-test".to_string(),
+            feature_type: None,
+            description: None,
+            created_at: None,
+            last_seen_at: None,
+            enabled: true,
+            stale: None,
+            impression_data: None,
+            project: None,
+            strategies: None,
+            variants: None,
+            dependencies: None,
+        },
+    };
+    delta_cache.add_events(vec![updated_feature_event]);
+
+    assert_eq!(delta_cache.get_events()[1], initial_feature_event);
```

Comment:
We'd expect 3 events in total here?
So `delta_cache.get_events()[2]` would be `updated_feature_event`?

Reply:
Yes, added that also to test.